src/udev/udev-watch.c

```diff
@@ -161,34 +165,351 @@ static int synthesize_change_one(sd_device *dev, sd_device *target) {
+                        log_device_debug(dev, "Partition UUID of '%s' does not match: blkid=%s, sysfs=%s",
+                                         node, p->uuid, uuid);
+                        return 0;
+                }
+        }
+
+        if (!hashmap_isempty(partitions)) {
+                Partition *p = hashmap_first(partitions);
+                log_device_debug(dev, "Partition '%s' found by blkid does not found in sysfs.", p->node);
                 return 0;
+        }
 
-        /* We have partitions but re-reading the partition table did not work, synthesize
-         * "change" for the disk and all partitions. */
-        r = synthesize_change_one(dev, dev);
-        FOREACH_DEVICE(e, d)
-                RET_GATHER(r, synthesize_change_one(dev, d));
+        if (partitions)
+                log_device_debug(dev, "All partitions found by blkid are available on sysfs, triggering 'change' events without requesting the kernel to reread partition table.");
+        else
+                log_device_debug(dev, "No partition found by blkid and on sysfs, triggering 'change' event for the disk without requesting the kernel to reread partition table.");
 
-        return r;
+        (void) synthesize_change_all(dev, e);
+        return 1; /* synced and sent change events */
+}
+
+static usec_t sync_partition_timeout_usec(void) {
+        static usec_t saved = 30 * USEC_PER_SEC;
+        static bool parsed = false;
+        int r;
+
+        if (parsed)
+                return saved;
+
+        parsed = true;
+
+        const char *e = getenv("SYSTEMD_UDEV_SYNC_PARTITION_TIMEOUT_SEC");
+        if (!e)
+                return saved;
+
+        usec_t timeout;
+        r = parse_sec(e, &timeout);
+        if (r < 0)
+                log_debug_errno(r, "Failed to parse $SYSTEMD_UDEV_SYNC_PARTITION_TIMEOUT_SEC=%s, ignoring: %m", e);
+
+        if (timeout > 5 * USEC_PER_HOUR) /* Add an arbitrary upper bound */
+                log_debug("Parsed $SYSTEMD_UDEV_SYNC_PARTITION_TIMEOUT_SEC=%s is too large, ignoring.", e);
+        else
+                saved = timeout;
+
+        return saved;
+}
+
+static int wait_for_partitions_synced(sd_device *dev) {
+        int r;
+
+        assert(dev);
+
+        usec_t timeout_usec = usec_add(now(CLOCK_MONOTONIC), sync_partition_timeout_usec());
+        usec_t range = 100 * USEC_PER_MSEC;
+
+        for (;;) {
+                r = partitions_synced(dev);
+                if (r < 0)
+                        return log_device_debug_errno(dev, r, "Failed to check if partitions are synced: %m");
+                if (r > 0)
+                        return 1; /* synced and sent change events */
+
+                usec_t now_usec = now(CLOCK_MONOTONIC);
+                if (now_usec >= timeout_usec)
+                        break;
+
+                (void) usleep_safe(MIN(usec_add(random_u64_range(range), 100 * USEC_PER_MSEC),
+                                       usec_sub_unsigned(timeout_usec, now_usec)));
```

Comment:
Is there something better we could listen for here? Rather than sleeping for an arbitrary amount of time, couldn't we have the kernel tell us about changes to the device / its partitions, with a timeout to fallback to rereading the table?

Reply:
I guess only interface is reading uevents. OK. I will try to do that.